### PR TITLE
Migrated codebase to work as an action 

### LINKED ISSUE #43: Migrate project to work as an action
## Why?
The idea of this change is to provide 3 improvements:
- Remove the need of a server.
- Greatly improve Developer Experience (DX) so anyone can use it.
- Rewrite the app to have a more consistent codebase.
### Remove the need of a server
Rewriting the GitHub app into an action would allow us to execute it as a single execution function, removing the need to have a running instance of this app in a server.
### Greatly improve DX
Currently, when someone decides to modify anything with the app, they must contact an API endpoint with a credential. This implementation is difficult which ends up requiring the OpsTooling to do it ourselves. So, when implementing this system we are the ones who need to install it. By converting it into an action other users can use this app without the need to have us intervene.
### Rewrite the app to have a more consistent codebase
The app has several entry points and connections, this document proposes a simplification of it. More info available in the architecture section.
### Entrypoint
The action's entrypoint will be set up by two kind of events:
```yml
on:
  issues:
    # updates when a issue is modified
    types:
      - opened
      - reopened
      - labeled
      - transferred
  # this can be manually triggered by going to the action's tab in the repository
  workflow_dispatch:
    inputs:
      syncIssues:
        description: 'Sync all issues available in the repository'
        required: true 
        type: boolean 
```
It will sync with two types of events, automatically, when a issue is modified, or manually, when the workflow is manually triggered by going to the action tab.
### Configuration
At the time of configurating rules we can set everything in the action file.
```yml
with:
  # Token with the following permissions: write:projects. 
  token: ${{ secrets.PROJECTS_TOKEN }}
  # The number of the project which the issues will be synced to
  project: 123
  # The name of the project field which the issue will be assigned to
  target-project-field: Team
  # The value which will be set in the field, in this case the team's name
  target-project-field-value: Foo
  # Optional field, only execute on issues with a given label
  label:
    - ops
    - ci
    - leads
```
For the token we would require one that has access only to project, allowing us to link the issue to a particular project. Some experimentation is required in this area to find the ideal permission access but it should be easy to do.
The token could be added either as per repository or a global one for [the organization](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-an-organization).
### Flow
The flow would be quite straightforward, if the action is initialized over a `workflow_dispatch`, it will fetch all the issues from the project and iterate over them assigning them the correct condition. 
But, if it's initialized through a Issue event, it will check which condition triggered the action and will act accordingly. You can visualize the flow in the following diagram:
```mermaid
flowchart LR
Start-up --> SyncOrIssue{Is it a\nworkflow dispatch?}
SyncOrIssue --> |Yes|Fetch[(Fetch all\nissues ID)]
Fetch --> Assign[[Assign Issue\nto Project Ids]]
SyncOrIssue --> |No|Switch((Switch over\nthe issue state))
Switch --> opened --> Assign
Switch --> transfered --> Assign
Switch --> labeled --> Label[[Calculate condition]] --> Assign
Switch --> closed --> ChangeState[[Change Issue\nState in Project]]
subgraph issue-event
opened
transfered
labeled
closed
end
```
### Architecture 
The architecture will be defined in an entry point (the core) and will initialize all the required values, sending everything to a business layer that will be the one in charge of executing the side effect.
```mermaid
flowchart LR
A(Event) --> |Action is initialized| Core(Core)
Core(Core) --> |Initialized GitHub app| Business(Business layer)
Business --> Logger(Logger)
Business --> Octokit(Octokit)
    subgraph Application
        Business
        Logger
        Octokit
    end
```
The core will parse all the configuration coming from both the event and the action workflow file and parse it into an agnostic object:
```typescript
// Example of the event data
interface IEventData {
	issueId?: number;
	event: "opened issue" | "closed issue" | "labeled issue" | "transfered issue" | "sync all issues"
	issueState?: "open" | "closed"
	config: {
		project: number,
		targetField: string,
		targetValue: string
	}
}
```
Everything related to interacting with GitHub (modifying the state, fetching configurations, etc) will be encapsulated in an wrapper that will be accessed by the business layer:
```typescript
interface IGitHub {
	assignProjectToIssue(issueId: number, projectId: number): Promise<boolean>;
	getIssueState(issueId: number): Promise<"open" | "closed">;
	getAllIssuesId(): Promise<number[]>;
	getProjectIdFromIssue(issueId: number): Promise<number>;
	changeIssueStateInProject(issueId: number, state: "todo" | "in progress" | "blocked" | "done");
}
```
Everything related to interacting with the logging system will also be wrapped in its own wrapper:
```typescript
interface ILogger {
	log(message: string): void;
	warn(message: string): void;
	error(message: string | Error): void;
}
```
This would allow us to have an agnostic business logic where we can write a clean and easy to understand logic.
```typescript
const octokit: IGitHub;
const logger: ILogger;
/** Overly simplified example of how it would work **/
async function execute(eventData: IEventData) {
    const { issueId, event } = eventData;
    if (event === "opened issue") {
        await octokit.assignProjectToIssue(eventData.issueId, eventData.config.project);
        logger.log(`Assigned issue #${issueId} to project ${eventData.config.project}!`);
    } else if (event === "closed issue") {
        if (await octokit.getProjectIdFromIssue(issueId) > 0) {
            await octokit.changeIssueStateInProject(issueId, "done");
            logger.log(`Updated issue #${issueId} to done!`);
        }
    } else if (event === "sync all issues") {
        const issues = await octokit.getAllIssuesId();
        for (let i = 0; i < issues.length; i++) {
            await octokit.assignProjectToIssue(issues[i], eventData.config.project);
            logger.log(`Assigned issue #${issues[i]} to project ${eventData.config.project}!`);
        }
    }
}
```

This would also **allow us to test the logic by substituting the dependencies**. And if, for whatever reason, we want to port this code to be a GitHub app or to run in GitLab, *the abstraction would make it quite straightforward*.
## Multiple project configuration
By setting up multiple executions, the action can be executed multiple times for different projects, allowing some better level of customization.
```yml
on:
  issues:
    # updates when a issue is modified
    types:
    - opened
    - reopened
    - labeled
    - transferred
  # this can be manually triggered by going to the action's tab in the repository
  workflow_dispatch:
    inputs:
      syncIssues:
        description: 'Sync all issues available in the repository'
        required: true 
        type: boolean 
jobs:
  lead-board:
    runs-on: ubuntu
    steps:
      - uses: paritytech/github-issue-sync
        with:
          token: ${{ secrets.PROJECTS_TOKEN }}
          project: 123
          label:
          - leads
  dev-board:
    runs-on: ubuntu
    steps:
      - uses: paritytech/github-issue-sync
        with:
          token: ${{ secrets.PROJECTS_TOKEN }}
          project: 321
          target-project-field: Team
          target-project-field-value: Foo
          label:
          - engineering
          - ci
```

# Completion requirements:
- [x] Works as an action.
- [x] Has an updated readme
- [x] Has unit tests.
- [x] Has been tested in some system.

### PR DESCRIPTION
**Pull request by Bullrich**
Updated codebase to run as an action instead of a application.

This PR mainly removes most of the boilerplate and code unrelated to the action itself and adapts everything to work in a more organized and abstract way.

It splits the architecture into an object oriented dependency injection system, having the class `Synchronizer` handling all the business logic and the classes `ProjectKit`,`IssueKit` and `CoreLogger` handling specific subtasks.

To be able to fine-grain the project permissions and keep the highest level of security possible while maintaining a sane codebase, the action of each token has been split into its own class:
- ProjectKit uses a PAT with organization access as it needs to interact with the project API (which is part of an organization instead of a repository).
- IssueKit uses the autogenerated GitHub token to fetch information from the issues API available in the repository.

`ProjectKit` is still in development, as it's missing the ability to update custom fields (#44) and it's using a deprecated API (#46) so it's prone to changes in the following PRs.

This PR is aimed at a feature branch and not at the master branch. The reason for this is because this task does not contemplate all the necessary features to launch this action. This is the first task of many which can be found in the milestone: [GitHub Action port](https://github.com/paritytech/github-issue-sync/milestone/1). Once all this tasks have been completed, the feature branch will be merged into master replacing the current functionality.

This PR also removes the ability to use custom labeling and custom fields. This will be added later to the feature branch.

The action was ported to work as a Docker image instead of a JavaScript image. This is because the docker image allows to build the action on demand (currently ~45 seconds). A more efficient method will be proposed later but for the current development process this satisfies the requirements.

This PR resolves #43.

### CODE REVIEW

**Comment by mutantcornholio** [comment]
> This PR is aimed at a feature branch and not at the master branch. The reason for this is because this task does not contemplate all the necessary features to launch this action. This is the first task of many which can be found in the milestone: [GitHub Action port](https://github.com/paritytech/github-issue-sync/milestone/1). Once all this tasks have been completed, the feature branch will be merged into master replacing the current functionality.  

Soo, should we also review the code in the feature branch then? Or was it reviewed already?

**Comment by Bullrich** [comment]
> Soo, should we also review the code in the feature branch then? Or was it reviewed already?

No, as there are no changes. The feature branch has just been created for this PR. It's currently up to date with master so there isn't anything new in it.

**Comment by mutantcornholio** [comment]
Ah, I was probably looking at the head branch, thinking it's the base...

File: .github/workflows/javascript-test.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Use node 18
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
```
Comment (mutantcornholio) [inline .github/workflows/javascript-test.yml:13]: 18 is already LTS, huh
guess we can start upgrading it in all the other places.

File: .github/workflows/javascript-test.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile
```
Comment (mutantcornholio) [inline .github/workflows/javascript-test.yml:15]: Coming back to you with this one :D https://github.com/paritytech/opstooling-testing/pull/1/files#r1071349775

Comment (Bullrich) [reply .github/workflows/javascript-test.yml]: Good catch. Time to update!

File: .github/workflows/javascript-test.yml (hunk 1 above)
Comment (Bullrich) [reply .github/workflows/javascript-test.yml:13]: Yup! Technically speaking we can even use the [native node test runner](https://nodejs.org/de/blog/announcements/v18-release-announce/#test-runner-module-experimental) for our projects but I needed some substitution library.

File: .prettierrc.cjs [hunk 3]
```diff
@@ -1 +0,0 @@
-module.exports = require("opstooling-js-style/src/prettier/configuration")
```
Comment (mutantcornholio) [inline .prettierrc.cjs:1]: I think we need this one back

File: .eslintrc.cjs [hunk 4]
```diff
@@ -1,26 +0,0 @@
...
-        ...tsConfOverride.rules,
-        "@typescript-eslint/explicit-module-boundary-types": "off",
-        "@typescript-eslint/no-unsafe-assignment": "off",
-        "@typescript-eslint/no-explicit-any": "off",
-        "@typescript-eslint/no-unsafe-argument": "off",
-        "@typescript-eslint/switch-exhaustiveness-check": "off",
-      },
-    },
-  ],
-};
```
Comment (mutantcornholio) [inline .eslintrc.cjs:26]: Same here, I think the codestyle config is still needed

File: .github/workflows/javascript-test.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    steps:
+      - uses: actions/checkout@v3
+      - name: Use node 18
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
+      - name: Install dependencies
+        run: yarn install --immutable
+      - name: Run tests
+        run: yarn test
```
Comment (mutantcornholio) [inline .github/workflows/javascript-test.yml:17]: Also, codestyle integration here is missing

**Review by mutantcornholio — CHANGES_REQUESTED** [review]

**Comment by mutantcornholio** [comment]
Other than that looks good

**Comment by Bullrich** [comment]
@mutantcornholio I'm planning on adding the linting rules again but in a different issue #47.

Can we wait until that PR?

**Comment by mutantcornholio** [comment]
> @mutantcornholio I'm planning on adding the linting rules again but in a different issue #47.
> 
> Can we wait until that PR?

Doesn't feel like something out of scope, but sure

**Review by mutantcornholio — APPROVED** [review]

**Pull request merged**